src/msg/async/Event.cc

```diff
@@ -39,14 +39,14 @@ class C_handle_notify : public EventCallback {
   C_handle_notify(EventCenter *c, CephContext *cc): center(c), cct(cc) {}
   void do_request(int fd_or_id) {
     char c[256];
+    int r = 0;
     do {
-      center->already_wakeup.set(0);
-      int r = read(fd_or_id, c, sizeof(c));
+      r = read(fd_or_id, c, sizeof(c));
       if (r < 0) {
-        ldout(cct, 1) << __func__ << " read notify pipe failed: " << cpp_strerror(errno) << dendl;
-        break;
+        if (errno != EAGAIN)
+          ldout(cct, 1) << __func__ << " read notify pipe failed: " << cpp_strerror(errno) << dendl;
       }
-    } while (center->already_wakeup.read());
+    } while (r > 0);
```

Comment:
iiuc, as long as `read() > 0` , we should break the loop and return. but your implementation will not stop until the other peer close the fd or something goes wrong.